prow/pod-utils/clone/clone.go

```diff
@@ -57,6 +57,11 @@ func Run(refs *kube.Refs, dir, gitUserName, gitUserEmail, cookiePath string, env
 	commands = append(commands, shellCloneCommand(cloneDir, env, "git", "fetch", repositoryURI, "--tags", "--prune"))
 	commands = append(commands, shellCloneCommand(cloneDir, env, "git", "fetch", repositoryURI, refs.BaseRef))
 
+	// unless the user specifically asks us not to, init submodules
+	if !refs.SkipSubmodules {
+		commands = append(commands, shellCloneCommand(cloneDir, env, "git", "submodule", "update", "--init", "--recursive"))
```

Comment:
I think the way to test `clone.go` would be given a set of inputs to `Run()` inputs, what is in the list of commands?

Probably involving some sort of refactoring to allow the test to replace the `for _, command := range commands { fc, o, err := command() ... }` loop with something that checks each item instead somehow.


I suspect this PR is not adding the incremental coverage for `if !refs.SkipSubModules` because we don't already have the infrastructure in place for testing the rest of the branches here.

Reply:
`clone.go`
```go
var commandRunner = func(cc cloneCommand) (string, string, error) {
  return cc()
}

...

	for _, command := range commands {
		formattedCommand, output, err := commandRunner(command())
```

`clone_test.go`
```go
func TestFoo(t *testing.T) {
  old := commandRunner
  defer func(){commandRunner = old}()
  var actual []cloneCommand
  commandRunner = func(cc cloneCommad) (string, string, error) {
    actual = append(actual, cc)
    if cc.something == "fail" { return "", "", errors.New("injected whatever error") }
    return "", "", nil
  }
  cases := []struct{ 
      name string
      refs *kube.Refs
      dir string
      gitUserName string
      gitUserEmail string
      cookiePath string
      env []string
      expectedCommands []cloneCommand
      expectedRecord Record
  }{ { ...}, ... }
  for _, tc := range cases { 
    record := Run(tc.refs, tc.dir, tc.gun, tc.gue, tc.cp, tc.e)
    if !fancyEqual(actual, tc.expectedCommands) { t.Errorf("whatever") }
  }
}
```

This will also require making `cloneCommand` into a type where we can actually compare what the commands are (can't compare functions)

Reply:
Yes, that's what I suggested as a test harness and due to the fact that none of it exists, was wondering if you wanted the harness in this PR or a follow-up.